frame/uniswap-v2/src/lib.rs

```diff
@@ -0,0 +1,857 @@
+			+ Parameter
+			+ Codec
+			+ Copy
+			+ Ord
+			+ CheckedAdd
+			+ CheckedSub
+			+ CheckedMul
+			+ AtLeast32BitUnsigned
+			+ From<u64> // at least 64 bit
+			+ Zero
+			+ One
+			+ IntegerSquareRoot
+			+ FixedPointOperand
+			+ Into<LiftedFixedBalance>
+			+ Into<u128>; // cannot do From<u128>, until LiftedFixedBalance integer part is larger than 128
+			  // bit
+		type CurrencyFactory: CurrencyFactory<<Self as Config>::AssetId>;
+		type LpToken: Transfer<Self::AccountId, Balance = Self::Balance, AssetId = <Self as Config>::AssetId>
+			+ Mutate<Self::AccountId, Balance = Self::Balance, AssetId = <Self as Config>::AssetId>
+			+ Inspect<Self::AccountId, Balance = Self::Balance, AssetId = <Self as Config>::AssetId>;
+		type Precision: Get<FixedU128>;
+		type PoolId: FullCodec
+			+ Default
+			+ TypeInfo
+			+ Eq
+			+ PartialEq
+			+ Ord
+			+ Copy
+			+ Debug
+			+ CheckedAdd
+			+ One;
+		type PoolTokenIndex: Copy + Debug + Eq + Into<u32>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// Current number of pools (also ID for the next created pool)
+	#[pallet::storage]
+	#[pallet::getter(fn pool_count)]
+	pub type PoolCount<T: Config> = StorageValue<_, T::PoolId, ValueQuery>;
+
+	/// Existing pools
+	#[pallet::storage]
+	#[pallet::getter(fn pools)]
+	pub type Pools<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::PoolId, ConstantProductPoolInfo<T::AccountId>>;
+
+	/// Pool's LP asset
+	#[pallet::storage]
+	#[pallet::getter(fn pool_lp_asset)]
+	pub type PoolLPAsset<T: Config> = StorageMap<_, Blake2_128Concat, T::PoolId, T::AssetId>;
+
+	/// List of assets supported by the pool
+	#[pallet::storage]
+	#[pallet::getter(fn pool_assets)]
+	pub type PoolAssets<T: Config> = StorageMap<_, Blake2_128Concat, T::PoolId, Vec<T::AssetId>>;
+
+	/// Balance of asset for given pool excluding admin_fee
+	#[pallet::storage]
+	#[pallet::getter(fn pool_asset_balance)]
+	pub type PoolAssetBalance<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::PoolId,
+		Blake2_128Concat,
+		T::AssetId,
+		T::Balance,
+		ValueQuery,
+	>;
+
+	/// Balance of asset for given pool including admin_fee
+	#[pallet::storage]
+	#[pallet::getter(fn pool_asset_total_balance)]
+	pub type PoolAssetTotalBalance<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::PoolId,
+		Blake2_128Concat,
+		T::AssetId,
+		T::Balance,
+		ValueQuery,
+	>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Could not create new asset
+		AssetNotCreated,
+		/// Values in the storage are inconsistent
+		InconsistentStorage,
+		/// Not enough assets provided
+		NotEnoughAssets,
+		/// Some provided assets are not unique
+		DuplicateAssets,
+		/// Pool with specified id is not found
+		PoolNotFound,
+		/// Error occurred while performing math calculations
+		Math,
+		/// Specified asset amount is wrong
+		WrongAssetAmount,
```

Comment:
It is not clear what is wrong, like is it not enough to do some operation or not balanced or it does not matches some exact amount of something? like error should be more clear about what is exactly wrong and guide what to do

Reply:
Based on its usage , it indicates that asset amounts should be grater than zero.

Reply:
I will update it.